docs/source/usage.md

```diff
@@ -474,56 +475,53 @@ hoistFirst :: (forall x. m x -> m x) -> Sequential m a -> Sequential m a
 
-`FreeSampler m` is best understood if you're familiar with the standard use of a free monad to construct a domain specific language. For probability in particular, see this [blog post](https://jtobin.io/simple-probabilistic-programming). Here's the definition:
+The former is relatively straightforward: the `MonadSample` instance implements `random` as `get`ting the trace (using `get` from `MonadState`), using (and removing) the first element (`put` from `MonadState`), and writing that element to the output (using `tell` from `MonadWriter`). If the trace is empty, the `random` from the underlying monad is used, but the result is still written with `tell`.
+
+The latter is best understood if you're familiar with the standard use of a free monad to construct a domain specific language. For probability in particular, see this [blog post](https://jtobin.io/simple-probabilistic-programming). Here's the definition:
 
 ```haskell
 newtype SamF a = Random (Double -> a)
-newtype FreeSampler m a = 
-    FreeSampler {runFreeSampler :: FT SamF m a}
+newtype Density m a = 
+    Density {density :: FT SamF m a}
     
-instance Monad m => MonadSample (FreeSampler m) where
-  random = FreeSampler $ liftF (Random id)
+instance Monad m => MonadSample (Density m) where
+  random = Density $ liftF (Random id)
 ```
 
 The monad-bayes implementation uses a more efficient implementation of `FreeT`, namely `FT` from the `free` package, known as the *Church transformed Free monad*. This is a technique explained in https://begriffs.com/posts/2016-02-04-difference-lists-and-codennsity.html. But that only changes the operational semantics - performance aside, it works just the same as the standard `FreeT` datatype. 
 
 If you unpack the definition, you get:
 
 ```haskell
-FreeSampler m a ~ m (Either a (Double -> (FreeSampler m a)))
+Density m a ~ m (Either a (Double -> (Density m a)))
 ```
 
-As you can see, this is rather like `Coroutine`, except to "resume", you must provide a new `Double`, corresponding to the value of some particular random choice.
-
 Since `FreeT` is a transformer, we can use `lift` to get a `MonadSample` instance.
 
-
-A *trace* of a program of type `MonadSample m => m a` is an execution of the program, so a choice for each of the random values. Recall that `random` underlies all of the random values in a program, so a trace for a program is fully specified by a list of `Double`s, giving the value of each call to `random`.
-
-Given a probabilistic program interpreted in `FreeSampler m`, we can "run" it to produce a program in the underlying monad `m`. For simplicity, consider the case of a program `bernoulli 0.5 :: FreeSampler SamplerIO Bool`. We can then use the following function:
+`density` is then defined using the canonical property of the free monad (transformer), embodied by `iterFT`, which interprets `SamF` in the appropriate way:
```

Comment:
I think the defining property of a free monad is the fold: https://hackage.haskell.org/package/free-5.1.9/docs/Control-Monad-Free.html#v:foldFree and https://hackage.haskell.org/package/free-5.1.9/docs/Control-Monad-Trans-Free.html#v:foldFreeT which means the free monad is initial (and hence can be interpreted in any desired way). TBH I don't know how this relates to `iterFT`. What does canonical mean here?

Reply:
Updated.